source/fab/steps/walk_source.py

```diff
@@ -0,0 +1,88 @@
+##############################################################################
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+##############################################################################
+"""
+Gather files from a source folder.
+
+"""
+import logging
+from pathlib import Path
+from typing import Optional, List, Tuple
+
+from fab.config import PathFilter
+from fab.constants import BUILD_OUTPUT
+from fab.steps import Step
+from fab.util import file_walk
+
+logger = logging.getLogger('fab')
+
+
+INCLUDE = True
+EXCLUDE = False
+
+
+class FindSourceFiles(Step):
+
+    def __init__(self,
+                 source_root: Path, output_name="all_source",
+                 build_output: Optional[Path] = None, name="Walk source",
+                 file_filtering: Optional[List[Tuple]] = None):
+        """
+        Args:
+            - source_root: Path to source folder.
+            - output_name: Name of artefact to create, defaults to "all_source".
+            - build_output: (Deprecated) where to create the output folders.
+            - file_filtering: List of (file pattern, boolean) tuples.
+                    Processed in order, if a source file matches the pattern it will be included or excluded,
+                    as per the bool.
+
+        """
+        super().__init__(name)
+        self.source_root = source_root
+        self.output_artefact = output_name
+        self.build_output = build_output or source_root.parent / BUILD_OUTPUT
+
+        file_filtering = file_filtering or []
+        self.path_filters: List[PathFilter] = [PathFilter(*i) for i in file_filtering]
+
+    def run(self, artefacts, config):
+        """
+        Get all files in the folder and subfolders.
+
+        Requires no artefacts, creates the "all_source" artefact.
+
+        """
+        super().run(artefacts, config)
+
+        fpaths = list(file_walk(self.source_root))
```

Comment:
We wanted it truthy to check if it's empty and raise an exception. However, since we've added the filtering, it's better we check there's files after filtering, so we'll just iterate through the generator again now.